service/controller/release/resource/argoapps/resource.go

```diff
@@ -0,0 +1,255 @@
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("calculated %d obsolete apps", len(appsToDelete.Items)))
+	for i, app := range appsToDelete.Items {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleting app %#q in namespace %#q", app.GetName(), app.GetNamespace()))
+
+		err := r.k8sClient.CtrlClient().Delete(
+			ctx,
+			&appsToDelete.Items[i],
+		)
+		if apierrors.IsNotFound(err) {
+			// fall through.
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleted app %#q in namespace %#q", app.GetName(), app.GetNamespace()))
+	}
+
+	appsToCreate := calculateMissingApps(componentApps, apps)
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("calculated %d missing apps", len(appsToCreate.Items)))
+	for i, app := range appsToCreate.Items {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("creating app %#q in namespace %#q", app.GetName(), app.GetNamespace()))
+
+		// ensure giantswarm.io/managed-by label
+		if appsToCreate.Items[i].Labels == nil {
+			appsToCreate.Items[i].Labels = map[string]string{}
+		}
+		appsToCreate.Items[i].Labels[key.LabelManagedBy] = project.Name()
+
+		err := r.k8sClient.CtrlClient().Create(
+			ctx,
+			&appsToCreate.Items[i],
+		)
+		if apierrors.IsAlreadyExists(err) {
+			// fall through.
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("created app %#q in namespace %#q", app.GetName(), app.GetNamespace()))
+	}
+
+	return nil
+}
+
+func calculateMissingApps(componentApps map[string]argoappv1alpha1.Application, apps argoappv1alpha1.ApplicationList) argoappv1alpha1.ApplicationList {
+	var missingApps argoappv1alpha1.ApplicationList
+
+	for _, componentApp := range componentApps {
+		missingApp := true
+		for _, app := range apps.Items {
+			if compareArgoApplications(componentApp, app) {
+				missingApp = false
+				break
+			}
+		}
+		if missingApp {
+			missingApps.Items = append(missingApps.Items, componentApp)
+		}
+	}
+
+	return missingApps
+}
+
+func calculateObsoleteApps(componentApps map[string]argoappv1alpha1.Application, apps argoappv1alpha1.ApplicationList) argoappv1alpha1.ApplicationList {
+	var obsoleteApps argoappv1alpha1.ApplicationList
+
+	for _, app := range apps.Items {
+		obsoleteApp := true
+		for _, componentApp := range componentApps {
+			if compareArgoApplications(componentApp, app) {
+				obsoleteApp = false
+				break
+			}
+		}
+		if obsoleteApp {
+			obsoleteApps.Items = append(obsoleteApps.Items, app)
+		}
+	}
+
+	return obsoleteApps
+}
+
+func compareArgoApplications(a, b argoappv1alpha1.Application) bool {
+	if a.Name != b.Name {
+		return false
+	}
+
+	konfigureA := extractKonfigureVariables(a)
+	konfigureB := extractKonfigureVariables(b)
+
+	match := true
+	match = match && konfigureA.AppName != "" && konfigureA.AppName == konfigureB.AppName
```

Comment:
Why can we not compare the `konfigureVariables` struct directly here? Or implement a method `func (r *konfigureVariables) compare(input konfigureVariables) bool`?

Reply:
Good advice, let me fix that.